lib/event_dispatcher/dispatcher.rb

```diff
@@ -0,0 +1,104 @@
+module EventDispatcher
+  class Dispatcher
+    private
+
+    attr_reader :event_listeners
+    attr_reader :sorted
+
+    public
+
+    def initialize
+      @event_listeners = {}
+      @sorted = {}
+    end
+
+    def dispatch(event_name:, event: EventDispatcher::BaseEvent.new)
+      listeners = listeners(event_name: event_name)
+
+      do_dispatch(listeners: listeners, event: event) if listeners
+
+      event
+    end
+
+    def listeners(event_name: nil)
+      unless event_name.nil?
+        return [] if !event_listeners.key?(event_name) || event_listeners[event_name].empty?
+
+        sort_listeners(event_name: event_name) if sorted[event_name].nil? || sorted[event_name].empty?
+
+        return sorted[event_name]
+      end
+
+      event_listeners.each_key do |key|
+        sort_listeners(event_name: key) if sorted[key].nil? || sorted[key].empty?
+      end
+
+      sorted
+    end
```

Comment:
I'm also a bit concerned with this method being `public`. For a `private` implementation, it's completely ok, but if it's part of the API, you're exposing the mutable listeners list/hash you use to the outside.